### PR TITLE
feat(JOML): migrate gltf and rendering

### PR DESCRIPTION
**Pull request by pollend**

### CODE REVIEW

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFAnimationFormat.java [hunk 1]
```diff
@@ -114,37 +117,41 @@ private MeshAnimationData loadAnimation(GLTF gltf, GLTFAnimation animation, List
...
-                    channelReaders.add(new BufferChannelReader<>(times, data, sampler.getInterpolation()::interpolate, x -> x.getRotation(bone)));
+                    List<Quaternionf> data = getQuat4fs(gltf, loadedBuffers, sampler.getOutput());
+                    channelReaders.add(new BufferChannelReader<>(times, data, sampler.getInterpolation()::interpolate
+                        , x -> x.getRotation(bone)));
                     break;
                 }
                 case SCALE: {
-                    List<Vector3f> data = getVector3fs(gltf, loadedBuffers, sampler.getOutput());
-                    channelReaders.add(new BufferChannelReader<>(times, data, sampler.getInterpolation()::interpolate, x -> x.getBoneScale(bone)));
+                    List<org.joml.Vector3f> data = getVector3fs(gltf, loadedBuffers, sampler.getOutput());
```
Comment (DarkWeird) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFAnimationFormat.java:133]: FQN of class not needed more (import was removed)

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFAnimationFormat.java [hunk 2]
```diff
@@ -114,37 +117,41 @@ private MeshAnimationData loadAnimation(GLTF gltf, GLTFAnimation animation, List
 
             switch (channel.getTarget().getPath()) {
                 case TRANSLATION: {
-                    List<Vector3f> data = getVector3fs(gltf, loadedBuffers, sampler.getOutput());
+                    List<org.joml.Vector3f> data = getVector3fs(gltf, loadedBuffers, sampler.getOutput());
```
Comment (DarkWeird) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFAnimationFormat.java:120]: FQN of class not needed more (import was removed)

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFAnimationFormat.java [hunk 3]
```diff
@@ -163,20 +170,20 @@ private TFloatList getFloats(GLTF gltf, List<byte[]> loadedBuffers, int accessor
         return floats;
     }
 
-    private List<Vector3f> getVector3fs(GLTF gltf, List<byte[]> loadedBuffers, int accessorIndex) throws IOException {
+    private List<org.joml.Vector3f> getVector3fs(GLTF gltf, List<byte[]> loadedBuffers, int accessorIndex) throws IOException {
```
Comment (DarkWeird) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFAnimationFormat.java:173]: FQN of class not needed more (import was removed)

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFCommonFormat.java [hunk 4]
```diff
@@ -249,7 +272,7 @@ protected void checkVersionSupported(ResourceUrn urn, GLTF gltf) throws IOExcept
             if (Strings.isNullOrEmpty(boneName)) {
                 boneName = "bone_" + i;
             }
-            Bone bone = new Bone(i, boneName, new Matrix4f(rotation, position, scale.x));
+            Bone bone = new Bone(i, boneName, new org.joml.Matrix4f().translationRotateScale(position, rotation, scale));
```
Comment (DarkWeird) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFCommonFormat.java:275]: FQN of class not needed more (import was removed)

File: engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java [hunk 5]
```diff
@@ -31,7 +30,7 @@
     private String name;
     private int index;
     private Matrix4f relativeTransform = new Matrix4f();
-    private Matrix4f inverseBindMatrix = new Matrix4f(Matrix4f.IDENTITY);
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:34]: This seems like those two matrices should be different... why do we joml-migrate both to a plain new matrix?

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFCommonFormat.java [hunk 6]
```diff
@@ -293,7 +316,7 @@ protected void checkVersionSupported(ResourceUrn urn, GLTF gltf) throws IOExcept
                     values.get(i + 8), values.get(i + 9), values.get(i + 10), values.get(i + 11),
                     values.get(i + 12), values.get(i + 13), values.get(i + 14), values.get(i + 15)
             );
-            mat.transpose();
+//            mat.transpose();
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFCommonFormat.java:319]: what about this? still needed or can be removed?

**Review by jdrueckert — CHANGES_REQUESTED** [review]

File: engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java (hunk 5 above)
Comment (pollend) [reply engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:34]: Matricies in JOML are always Identity.

Comment (jdrueckert) [reply engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:34]: yes, but `Matrix4f()` wasn't, was it?

File: engine/src/main/java/org/terasology/rendering/logic/SkeletonRenderer.java [hunk 7]
```diff
@@ -288,20 +288,16 @@ public void renderOpaque() {
...
 
                 }
             }
             ((OpenGLSkeletalMesh) skeletalMesh.mesh).setScaleTranslate(skeletalMesh.scale, skeletalMesh.translate);
-            ((OpenGLSkeletalMesh) skeletalMesh.mesh).render(Arrays.stream(boneTransforms).map(k -> {
-                org.terasology.math.geom.Matrix4f t = JomlUtil.from(k);
-                t.transpose();
-                return t;
-            }).collect(Collectors.toList()));
+            ((OpenGLSkeletalMesh) skeletalMesh.mesh).render(Arrays.stream(boneTransforms).collect(Collectors.toList()));
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/logic/SkeletonRenderer.java:300]: No need to use a Stream here:
```suggestion
            ((OpenGLSkeletalMesh) skeletalMesh.mesh).render(Arrays.asList(boneTransforms));
```

File: engine/src/main/java/org/terasology/rendering/gltf/model/GLTFInterpolation.java [hunk 8]
```diff
@@ -30,15 +30,17 @@
     LINEAR {
         @Override
         public void interpolate(Vector3f a, Vector3f b, float t, Vector3f out) {
-            out.x = a.x + t * (b.x - a.x);
-            out.y = a.y + t * (b.y - a.y);
-            out.z = a.z + t * (b.z - a.z);
+            a.lerp(b,t,out);
+//            out.x = a.x + t * (b.x - a.x);
+//            out.y = a.y + t * (b.y - a.y);
+//            out.z = a.z + t * (b.z - a.z);
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/model/GLTFInterpolation.java:36]: remove commented code and apply formatter.
```suggestion
            a.lerp(b, t, out);
```

File: engine/src/main/java/org/terasology/rendering/gltf/model/GLTFInterpolation.java [hunk 9]
```diff
@@ -30,15 +30,17 @@
...
 
         }
 
         @Override
-        public void interpolate(Quat4f a, Quat4f b, float t, Quat4f out) {
-            out.set(BaseQuat4f.interpolate(a, b, t));
+        public void interpolate(Quaternionf a, Quaternionf b, float t, Quaternionf out) {
+            a.slerp(b, t, out);
+//            out.set(BaseQuat4f.interpolate(a, b, t));
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/model/GLTFInterpolation.java:43]: remove commented code
```suggestion
```

File: engine/src/main/java/org/terasology/rendering/gltf/deserializers/Vector3fDeserializer.java [hunk 10]
```diff
@@ -21,7 +21,7 @@
 import com.google.gson.JsonParseException;
 import gnu.trove.list.TFloatList;
 import gnu.trove.list.array.TFloatArrayList;
-import org.terasology.math.geom.Vector3f;
+import org.joml.Vector3f;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/deserializers/Vector3fDeserializer.java:24]: Ehm, what is this deserializer used for? 

I would assume we keep all de- and encoding logic for the TeraMath data types and add new ones for the JOML types (if needed). Are we sure this should be changed?

File: engine/src/main/java/org/terasology/rendering/gltf/deserializers/Quat4fDeserializer.java [hunk 11]
```diff
@@ -21,21 +21,21 @@
...
-import org.terasology.math.geom.Quat4f;
+import org.joml.Quaternionf;
 
 import java.lang.reflect.Type;
 
 /**
  * Json deserializer for an Quat4f.
  */
-public class Quat4fDeserializer implements JsonDeserializer<Quat4f> {
+public class Quat4fDeserializer implements JsonDeserializer<Quaternionf> {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/deserializers/Quat4fDeserializer.java:31]: Same here, should we keep the old deserializer for teramath types?

File: engine/src/main/java/org/terasology/rendering/gltf/deserializers/Matrix4fDeserializer.java [hunk 12]
```diff
@@ -21,7 +21,7 @@
 import com.google.gson.JsonParseException;
 import gnu.trove.list.TFloatList;
 import gnu.trove.list.array.TFloatArrayList;
-import org.terasology.math.geom.Matrix4f;
+import org.joml.Matrix4f;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/deserializers/Matrix4fDeserializer.java:24]: Same here, should we keep the old deserializer for teramath types?

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFMeshFormat.java [hunk 13]
```diff
@@ -131,19 +131,19 @@ private Matrix4f getMatrix(GLTF gltf, int nodeIndex) {
...
-                transform.set(0, 2, scale.getX() * transform.get(0, 2));
-
-                transform.set(1, 0, scale.getY() * transform.get(1, 0));
-                transform.set(1, 1, scale.getY() * transform.get(1, 1));
-                transform.set(1, 2, scale.getY() * transform.get(1, 2));
-
-                transform.set(2, 0, scale.getZ() * transform.get(2, 0));
-                transform.set(2, 1, scale.getZ() * transform.get(2, 1));
-                transform.set(2, 2, scale.getZ() * transform.get(2, 2));
+                transform.set(new Matrix4f().translationRotateScale(position,rotation, scale));
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFMeshFormat.java:134]: Why do we create a new matrix object here and set `transform`? Isn't this equivalent to just
```suggestion
                transform.translationRotateScale(position,rotation, scale);
```

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFMeshFormat.java [hunk 14]
```diff
@@ -131,19 +131,19 @@ private Matrix4f getMatrix(GLTF gltf, int nodeIndex) {
...
+//                transform.set(0, 1, scale.getX() * transform.get(0, 1));
+//                transform.set(0, 2, scale.getX() * transform.get(0, 2));
+//
+//                transform.set(1, 0, scale.getY() * transform.get(1, 0));
+//                transform.set(1, 1, scale.getY() * transform.get(1, 1));
+//                transform.set(1, 2, scale.getY() * transform.get(1, 2));
+//
+//                transform.set(2, 0, scale.getZ() * transform.get(2, 0));
+//                transform.set(2, 1, scale.getZ() * transform.get(2, 1));
+//                transform.set(2, 2, scale.getZ() * transform.get(2, 2));
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFMeshFormat.java:146]: remove commented code
```suggestion
```

File: engine/src/main/java/org/terasology/rendering/gltf/GLTFCommonFormat.java [hunk 15]
```diff
@@ -65,7 +88,7 @@
             .registerTypeAdapter(TIntList.class, new TIntListDeserializer())
             .registerTypeAdapter(TFloatList.class, new TFloatListDeserializer())
             .registerTypeAdapter(Matrix4f.class, new Matrix4fDeserializer())
-            .registerTypeAdapter(Quat4f.class, new Quat4fDeserializer())
+            .registerTypeAdapter(Quaternionf.class, new Quat4fDeserializer())
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/gltf/GLTFCommonFormat.java:91]: I'm not sure whether we should change the deserializers for the old types in place, or phase them out more gracefully. Are they used anywhere else, or just here for the glTF format?

File: engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java [hunk 16]
```diff
@@ -92,32 +91,18 @@ public void addChild(Bone child) {
...
-        descaled.m21 /= scale.y;
-        descaled.m02 /= scale.z;
-        descaled.m12 /= scale.z;
-        descaled.m22 /= scale.z;
-        Quat4f result = new Quat4f(Quat4f.IDENTITY);
-        result.set(descaled);
+    public Quaternionf getLocalRotation() {
+        Quaternionf result = new Quaternionf();
+        relativeTransform.getNormalizedRotation(result);
         return result;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:101]:
```suggestion
        return relativeTransform.getNormalizedRotation(new Quaternionf());
```

**Review by skaldarnar — CHANGES_REQUESTED** [review]
The models in JS and MR look and behave the same as before. 

This does **not** fix the shadow bug on models, though.

I've requested changes to clean up the code a bit before we merge this.

**Edit**: I also see a couple of these `INFO` logs. Are they new?
```log
14:06:02.904 [main] INFO  o.t.logic.behavior.core.ActionNode - Exception while running construct() of action org.terasology.behaviors.actions.SetAnimationAction@1b03cf96 from entity EntityRef{id = 8983, netId = 0, prefab = 'MetalRenegades:enemyGooey'}: bound must be positive
```

File: engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java (hunk 5 above)
Comment (pollend) [reply engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:34]: it needs to be a zero matrix. umm I don't think there is a way to declare an empty matrix in joml.

File: engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java [hunk 17]
```diff
@@ -97,8 +97,7 @@ public Vector3f getLocalPosition() {
 
     public Quaternionf getLocalRotation() {
         Quaternionf result = new Quaternionf();
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:99]:
```suggestion
```

**Review by skaldarnar — CHANGES_REQUESTED** [review]

File: engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java (hunk 5 above)
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/rendering/assets/skeletalmesh/Bone.java:34]: The initial value is overwritten in the constructor anyways, so this is no problem.

**Review by skaldarnar — APPROVED** [review]
The info log that reads like a warning is pre-existing, and everything seems to work fine. Thus, LGTM!

**Pull request merged**